File: Lecture 1/problem_sets/credit/credit.c
#include <cs50.h>
#include <stdio.h>


int checksum(long number);

int main(void)
{
    do
    {
        long credit_card = get_long("Number: ");

        if (checksum(credit_card))
        {
            if (credit_card >= 1000000000000 && credit_card < 10000000000000) // Might be VISA.  13 digit VISA
            {
                int first_dig = credit_card / 1000000000000;
                if (first_dig == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }


            else if (credit_card >= 100000000000000 && credit_card < 1000000000000000) // Might be AMEX 15 digit
            {
                int first_two = credit_card / 10000000000000;

                if (first_two == 34 || first_two == 37)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }

            else if (credit_card >= 1000000000000000 && credit_card < 10000000000000000) // 16 digits
            {
                int first_dig = credit_card / 1000000000000000;
                int first_two = credit_card / 100000000000000;

                if (first_dig == 4)
                {
                    printf("VISA\n");
                }
                else if (first_two == 51 || first_two == 52 || first_two == 53 || first_two == 54 || first_two == 55)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n");
            }
            // 2221000000000009

        }
        else
        {
            printf("INVALID\n");
        }



    }
    while (false);

}



//might return bool but don't want)
int checksum(long number)
{

    // taking every digit

    int first = number % 10, second = (number / 10) % 10, third = (number / 100) % 10, fourth = (number / 1000) % 10,
        fifth = (number / 10000) % 10, sixth = (number / 100000) % 10, seventh = (number / 1000000) % 10, eighth = (number / 10000000) % 10,
        ninth = (number / 100000000) % 10, tenth = (number / 1000000000) % 10, eleventh = (number / 10000000000) % 10,
        twelveth = (number / 100000000000) % 10, thirteenth = (number / 1000000000000) % 10, fourteenth = (number / 10000000000000) % 10,
        fifteenth = (number / 100000000000000) % 10, sixteenth = (number / 1000000000000000) % 10;

    // as soon as there shouldn't be more than 9 and 9 * 2 = 18 maximum number is 18, so we can take every numbers modulo 10.
    int sum = (((second * 2) % 10) + first + (second * 2) / 10) + third + (((fourth * 2) % 10) + fifth +
              (fourth * 2) / 10) + seventh + (((sixth * 2) % 10) + ninth + (sixth * 2) / 10) + eleventh + (((eighth * 2) % 10) + thirteenth +
                      (eighth * 2) / 10) + fifteenth + (((tenth * 2) % 10) + (tenth * 2) / 10) + (((twelveth * 2) % 10) + (twelveth * 2) / 10) + (((
                                  fourteenth * 2) % 10) + (fourteenth * 2) / 10) + (((sixteenth * 2) % 10) + (sixteenth * 2) / 10);

    if (sum % 10 == 0)
    {
        //boolean true
        return 1;
    }
    else
    {
        //boolean false
        return 0;
    }

}

File: Lecture 6/sentimental-credit/credit.py
import cs50
import sys

def get_digits_sum(num):
    length = len(str(num))
    sum = 0
    for digit in str(num):
        sum += int(digit)

    return sum

def checSum(number):
    counter = 0
    for i in range(len(str(number)) - 2, -1, -2):

        if len(str(int(str(number)[i]) * 2)) > 1:
            sum = get_digits_sum(int(str(number)[i]) * 2)
            counter += sum + int(str(number)[i + 1])
        else:
            counter += (int(str(number)[i]) * 2) + int(str(number)[i + 1])


    if len(str(number)) % 2 != 0:
        counter += int(str(number)[0])

    return counter % 10 == 0

def main():

    credit_card_number = cs50.get_int("Number: ")

    if len(str(credit_card_number)) < 13 or len(str(credit_card_number)) > 16:
        print("INVALID")
        sys.exit(1)

    if checSum(credit_card_number):
        if str(credit_card_number)[0] == '4':
            print("VISA")
            sys.exit(0)
        elif str(credit_card_number)[:2] in ("51", "52", "53", "54", "55"):
            print("MASTERCARD")
            sys.exit(0)
        elif str(credit_card_number)[:2] in ("34", "37"):
            print("AMEX")
            sys.exit(0)
        else:
            print("INVALID")
            sys.exit(1)
    else:
        print("INVALID")
        sys.exit(1)

if __name__ == "__main__":
    main()